doc/src/devdocs/types.md

```diff
@@ -315,7 +315,129 @@ Ptr{Void} @0x00007f5998a04370
+`Tuple{T,T} where T<:Real`.
+However, `Tuple{Real}` *is* a subtype of `Tuple{T} where T`, because in that case
+`T` occurs only once and so is not diagonal.
+
+Next consider a signature like the following:
+
+```julia
+f(a::Array{T}, x::T, y::T) where T = ...
+```
+
+In this case, `T` occurs in invariant position inside `Array{T}`.
+That means whatever type of array is passed unambiguously determines
+the value of `T` --- we say `T` has an *equality constraint* on it.
+Therefore in this case the diagonal rule is not really necessary, since
+the array determines `T` and we can then allow `x` and `y` to be of
+any subtypes of `T`.
+So variables that occur in invariant position are never considered diagonal.
+This choice of behavior is slightly controversial, but your humble documentation
+author (Jeff in this case) likes it.
+
+The next complication is the interaction of unions and diagonal variables, e.g.
+
+```julia
+f(x::Union{Void,T}, y::T) where T = ...
+```
+
+Consider what this declaration means.
+`y` has type `T`. `x` then can have either the same type `T`, or else be of type `Void`.
+So all of the following calls should match:
+
+```julia
+f(1, 1)
+f("", "")
+f(2.0, 2.0)
+f(nothing, 1)
+f(nothing, "")
+f(nothing, 2.0)
+```
+
+These examples are telling us something: when `x` is `nothing::Void`, there are no
+extra constraints on `y`.
+It is as if the method signature had `y::Any`.
+This means that whether a variable is diagonal is not a static property based on
+where it appears in a type.
+Rather, it depends on where a variable appears when the subtyping algorithm *uses* it.
+When `x` has type `Void`, we don't need to use the `T` in `Union{Void,T}`, so `T`
+does not "occur".
+Indeed, we have the following type equivalence:
+
+```julia
+(Tuple{Union{Void,T},T} where T) == Union{Tuple{Void,Any}, Tuple{T,T} where T}
+```
+
+## Subtyping diagonal variables
+
+The subtyping algorithm for diagonal variables has two components:
+(1) identifying variable occurrences, and (2) ensuring that diagonal
+variables range over concrete types only.
+
+The first task is accomplished by keeping counters `occurs_inv` and `occurs_cov`
+(in `src/subtype.c`) for each variable in the environment.
+A variable is diagonal when `!occurs_inv && occurs_cov > 1`.
+
+The second task is accomplished by imposing a condition on a variable's lower bound.
+As the subtyping algorithm runs, it narrows the bounds of each variable
+(raising lower bounds and lowering upper bounds) to keep track of the
+range of variable values for which the subtype relation would hold.
+When we are done evaluating the body of a `UnionAll` type whose variable is diagonal,
+we look at the final values of the bounds.
+Since the variable must be concrete, a contradiction occurs if its lower bound is
+*not* concrete: while a concrete type can be a *subtype* of many other types, it is not
+possible for a concrete type to be a *supertype* of an abstract type, so the algorithm
```

Comment:
Are we excluding types like `Union{}` and `Type{Int}` from the definition of "abstract"?